### PR TITLE
Federated cmf

### PR DESCRIPTION
**Pull request by varkha-d-sharma**

### CODE REVIEW

File: cmflib/commands/init/amazonS3.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+        if result.stdout:
+            print(result.stdout)
+        file = "dvc_script_amazonS3.sh"
+        # find absolute path for dvc_script_amazonS3.sh
+        for root, dirs, files in os.walk(os.path.dirname(__file__)):
+            for name in files:
+                if name == file:
+                    abs_path = os.path.abspath(os.path.join(root, name))
+        # executing dvc_script_amazonS3.sh
+        result = subprocess.run(
```
Comment (annmary-roy) [inline cmflib/commands/init/amazonS3.py:58]: same here , check abs_path not none before proceeding

File: cmflib/commands/init/local.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+                create_cmf_config(cmf_config, "http://127.0.0.1:80")
+        abs_path = None
+        file = "git_initialize.sh"
+        # finding absolute path for git_initialize.sh
+        for root, dirs, files in os.walk(os.path.dirname(__file__)):
+            for name in files:
+                if name == file:
+                    abs_path = os.path.abspath(os.path.join(root, name))
+        # executing git_initialize.sh
+        result = subprocess.run(
```
Comment (annmary-roy) [inline cmflib/commands/init/local.py:44]: Take care of all the places for value of abs_path

File: cmflib/commands/init/show.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+from cmflib import cmfquery
+from cmflib.cli.command import CmdBase
+from cmflib.cli.utils import read_cmf_config, find_root
+
+
+class CmdInitShow(CmdBase):
+    def run(self):
+        msg = "'cmf' is not configured.\nExecute 'cmf init' command."
+        result = subprocess.run(
+            ["dvc", "config", "-l"],
```
Comment (annmary-roy) [inline cmflib/commands/init/show.py:32]: Handline of subprocess run should be made robust everywhere

File: cmflib/cli/utils.py [hunk 4]
```diff
@@ -1,4 +1,71 @@
...
+    msg = "'cmf' is not configured.\nExecute 'cmf init' command."
+    root_dir = os.path.realpath(os.getcwd())
+    while True:
+        file_path = os.path.join(root_dir, file_name)
+        if os.path.exists(file_path):
+            return root_dir
+        if os.path.ismount(root_dir):
+            return msg
+        parent = os.path.abspath(os.path.join(root_dir, os.pardir))
+        if parent == root_dir:
```
Comment (annmary-roy) [inline cmflib/cli/utils.py:36]: what is the scenario for this if statement ?

File: cmflib/merger.py [hunk 5]
```diff
@@ -1,73 +1,292 @@
...
-def parse_json_to_mlmd(mlmd_json):
+# mlmd is created from metadata passed in Json format
+def parse_json_to_mlmd(mlmd_json, path_to_store, cmd):
     mlmd_data = json.loads(mlmd_json)
-    type(mlmd_data)
+    # type(mlmd_data)
     pipelines = mlmd_data["Pipeline"]
-    print(type(pipelines))
+    # print(type(pipelines))
     pipeline = pipelines[0]
```
Comment (annmary-roy) [inline cmflib/merger.py:26]: This way we are restricting to a single pipeline. Although we expect only a single pipeline, still we can put a for loop here and make the code more generic to iterate through all the pipelines that are received

File: cmflib/commands/metadata/pull.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+        cmd = "pull"
+        mlmd_data = ""
+        status = 0
+        execution_flag = 0
+        if self.args.file_path:  # setting directory where mlmd file will be dumped
+            directory_to_dump = self.args.file_path
+        else:
+            directory_to_dump = os.getcwd()
+        output = server_interface.call_mlmd_pull(
+            url, self.args.pipeline_name
```
Comment (annmary-roy) [inline cmflib/commands/metadata/pull.py:45]: Here if execution id is given , we should pull only for that execution id, not the entire data. We need to be efficient in data transfer. So we cannot pull the entire data

File: cmflib/merger.py [hunk 7]
```diff
@@ -1,73 +1,292 @@
...
+    # print(type(pipeline))
+    pipeline_name = pipeline["name"]
+    # print(type(pipeline_name))
+    execution = {}
+    _ = None
+    cmf_class = cmf.Cmf(filename=path_to_store, pipeline_name=pipeline_name)
+    for stage in mlmd_data["Pipeline"][0]["stages"]:
+        # print(stage['name'])
+        for execution in stage["executions"]:
+            if execution["id"] == int(exec_id):
```
Comment (annmary-roy) [inline cmflib/merger.py:109]: After the execution id that we are interested is found, break from the loop

File: server/app/mlmd.py [hunk 8]
```diff
@@ -1,5 +1,10 @@
...
 
-def merge_mlmd(json_payload):
-    merger.parse_json_to_mlmd(json_payload)
+def merge_mlmd(json_payload,exec_id):
+    cmd='push'
+
+    if exec_id == None:                                                 #if execution id not available
+        merger.parse_json_to_mlmd(json_payload,'data/mlmd',cmd)         #parsing mlmd Json data to mlmd file
+    else:
+        merger.push_execution_to_mlmd(json_payload,'data/mlmd',cmd,exec_id) #parsing mlmd Json data for given execution id to mlmd file
```
Comment (annmary-roy) [inline server/app/mlmd.py:9]: ideally this should be a single function. There is lot repition in the code- for both functions. combine to single function. Other wise we will end up fixing bugs in both places

File: cmflib/cli/utils.py [hunk 9]
```diff
@@ -1,4 +1,71 @@
...
+def read_cmf_config(file_name: str):
+    with open(file_name, "r") as file:
+         return file.read()
+
+def git_exists():
+    try:
+        output = subprocess.check_output(["git", "version"]).decode("ascii")
+        return output
+    except FileNotFoundError:
+        return "\033[1;31mERROR:\033[1;m git is not installed!! Install git."
```
[the remark this answers is not in the document]
Comment (varkha-d-sharma) [reply cmflib/cli/utils.py:23]: These characters were introduced to print string "ERROR" in red color. Removed these characters for now. Will create a class which will handle color formatting.

File: cmflib/amazonS3_artifacts.py [hunk 10]
```diff
@@ -0,0 +1,52 @@
+###
+# Copyright (2022) Hewlett Packard Enterprise Development LP
+#
```
[the remark this answers is not in the document]
Comment (varkha-d-sharma) [reply cmflib/amazonS3_artifacts.py:3]: Created a new folder **storage_backends** and pushed the above mentioned files in the same folder.

[the remark this answers is not in the document]
Comment (varkha-d-sharma) [reply cmflib/amazonS3_artifacts.py:3]: Done

File: cmflib/dvc_config.py [hunk 11]
```diff
@@ -0,0 +1,72 @@
+###
+# Copyright (2022) Hewlett Packard Enterprise Development LP
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# You may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
```
[the remark this answers is not in the document]
Comment (varkha-d-sharma) [reply cmflib/dvc_config.py:8]: Created a new folder "utils" and pushed the file in the same folder.

File: cmflib/dvc_config.py [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+###
+
+import os
+import subprocess
+
+
+class dvc_config:
```
[the remark this answers is not in the document]
Comment (varkha-d-sharma) [reply cmflib/dvc_config.py:21]: Done.

File: cmflib/commands/artifact/pull.py [hunk 13]
```diff
@@ -42,49 +70,100 @@ def run(self):
...
         # print(names_urls)
-        artifact_class_obj = minio_artifacts.minio_artifacts()
-        for name_url in names_urls:
-            temp = name_url[1].split("/")
-            bucket_name = temp[2]
-            object_name = temp[3] + "/" + temp[4]
-            path_name = current_directory + "/" + name_url[0]
-            stmt = artifact_class_obj.download_artifacts(current_directory, bucket_name, object_name, path_name)
-            print(stmt)
+        dvc_config_op = dvc_config.get_dvc_config()  # pulling dvc config
```
[the remark this answers is not in the document]
Comment (varkha-d-sharma) [reply cmflib/commands/artifact/pull.py:79]: Done

File: cmflib/commands/artifact/pull.py [hunk 14]
```diff
@@ -16,24 +38,30 @@ def run(self):
             mlmd_file_name = self.args.file_name
             current_directory = os.path.dirname(self.args.file_name)
         if not os.path.exists(mlmd_file_name):
-            return f"{mlmd_file_name} doesn't exists in current directory"
+            return f"\033[1;31mERROR:\033[1;m {mlmd_file_name} doesn't exists in current directory."
```
[the remark this answers is not in the document]
Comment (varkha-d-sharma) [reply cmflib/commands/artifact/pull.py:41]: Changed the statement and removed extra special characters.

File: cmflib/commands/artifact/pull.py [hunk 15]
```diff
@@ -16,24 +38,30 @@ def run(self):
...
 
         identifiers = []
         for i in stages:
-            executions = query.get_all_executions_in_stage(i)  # getting all executions for stages
+            executions = query.get_all_executions_in_stage(
+                i
+            )  # getting all executions for stages
             dict_executions = executions.to_dict("dict")  # converting it to dictionary
-            identifiers.append(dict_executions['id'][0])  # id's of execution
+            identifiers.append(dict_executions["id"][0])  # id's of execution
```
[the remark this answers is not in the document]
Comment (abhinavchobey) [reply cmflib/commands/artifact/pull.py:51]: Done

File: cmflib/commands/artifact/pull.py [hunk 16]
```diff
@@ -16,24 +38,30 @@ def run(self):
...
```
[the remark this answers is not in the document]
Comment (abhinavchobey) [reply cmflib/commands/artifact/pull.py:52]: dict_executions dictionary format is dict_executions = {'id': {0:1}} that is why we used [0]

File: cmflib/commands/artifact/pull.py [hunk 17]
```diff
@@ -42,49 +70,100 @@ def run(self):
             url_list_updated = url_list_updated + url[i]
```
[the remark this answers is not in the document]
Comment (abhinavchobey) [reply cmflib/commands/artifact/pull.py:70]: url_list_updated has lists of url for each stage, so the step merges all the list and creates single list of url.

File: server/app/main.py [hunk 18]
```diff
@@ -1,21 +1,84 @@
...
+        if pipeline_name in query.get_pipeline_names():     #checks if pipeline name is available in mlmd
+            json_payload = query.dumptojson(pipeline_name)
+        else:
+            json_payload='NULL'
+    else:
+        print("No mlmd file submitted.")
+        json_payload=""
+    return json_payload
+
+# @app.get("/mlmd_pull_exec_by_id",response_class=HTMLResponse)
```
[the remark this answers is not in the document]
Comment (abhinavchobey) [reply server/app/main.py:47]: Removed the commented code, it was created for future use.

File: server/app/main.py [hunk 19]
```diff
@@ -1,21 +1,84 @@
...
 app = FastAPI()
 
+BASE_PATH=Path(__file__).resolve().parent
+templates=Jinja2Templates(directory=str(BASE_PATH/'template'))
 
 @app.get("/")
 def read_root():
-    return {"Use this api for sending json payload":"http://127.0.0.1:8000/mlmd_push/"}
-
+    return {"Use this api for sending json payload":"http://127.0.0.1/mlmd_push/"}
```
[the remark this answers is not in the document]
Comment (abhinavchobey) [reply server/app/main.py:19]: Done

File: server/app/main.py [hunk 20]
```diff
@@ -1,21 +1,84 @@
...
 
+#api to post mlmd file to cmf-server
 @app.post("/mlmd_push")
 async def mlmd_push(info: Request):
     req_info=await info.json()
-    merge_mlmd(req_info)
+    merge_mlmd(req_info['json_payload'],req_info['id'])
```
[the remark this answers is not in the document]
Comment (abhinavchobey) [reply server/app/main.py:26]: Here [id] means execution id, it has two values None or exec_id given by the user.

File: server/app/main.py [hunk 21]
```diff
@@ -1,21 +1,84 @@
...
+#     else:
+#         print("No mlmd file submitted.")
+#         json_payload=""
+#     return json_payload
+
+#api to display executions available in mlmd
+@app.get("/display_executions",response_class=HTMLResponse)
+async def display_exec(request: Request):
+    #checks if mlmd file exists on server
+    if os.path.exists("/cmf-server/data/mlmd"):
```
[the remark this answers is not in the document]
Comment (abhinavchobey) [reply server/app/main.py:63]: Done

File: server/app/get_data.py [hunk 22]
```diff
@@ -0,0 +1,39 @@
+from cmflib import cmfquery
+import pandas as pd
+
+#This function fetches all the executions available in given mlmd
+def index(mlmdfilepath):
```
[the remark this answers is not in the document]
Comment (abhinavchobey) [reply server/app/get_data.py:5]: Done

File: server/app/get_data.py [hunk 23]
```diff
@@ -0,0 +1,39 @@
...
+            # df = pd.concat([df, executions], sort=True, ignore_index=True)
+
+    name = []
+    url = []
+    df = pd.DataFrame()
+    for identifier in identifiers:
+        get_artifacts = query.get_all_artifacts_for_execution(identifier)  # getting all artifacts
+        artifacts_dict = get_artifacts.to_dict('dict')  # converting it to dictionary
+        df = pd.concat([df, get_artifacts], sort=True, ignore_index=True)
+    df.drop(['Commit','avg_prec','git_repo','last_update_time_since_epoch','metrics_name','model_framework','model_name','model_type','roc_auc','user-metadata1'],axis=1,inplace=True)
```
[the remark this answers is not in the document]
Comment (abhinavchobey) [reply server/app/get_data.py:36]: Done

File: cmflib/request_mlmdserver/server_interface.py [hunk 24]
```diff
@@ -1,9 +1,16 @@
 import requests
 import json
```
[the remark this answers is not in the document]
Comment (abhinavchobey) [reply cmflib/request_mlmdserver/server_interface.py:2]: Done

File: cmflib/merger.py (hunk 7 above)
Comment (abhinavchobey) [reply cmflib/merger.py:109]: Done

File: cmflib/commands/artifact/push.py [hunk 25]
```diff
@@ -0,0 +1,36 @@
...
+from cmflib.cli.command import CmdBase
+from cmflib.cli.utils import check_minio_server
+from cmflib.dvc_config import dvc_config
+
+
+class CmdArtifactPush(CmdBase):
+    def run(self):
+        dvc_config_op = dvc_config.get_dvc_config()
+        if(dvc_config_op[0] == "minio"):
+            if(check_minio_server() == "SUCCESS"):
```
[the remark this answers is not in the document]
Comment (varkha-d-sharma) [reply cmflib/commands/artifact/push.py:15]: Done

File: cmflib/cli/utils.py (hunk 4 above)
Comment (varkha-d-sharma) [reply cmflib/cli/utils.py]: It was not needed. Removed.

File: cmflib/commands/init/amazonS3.py (hunk 1 above)
Comment (varkha-d-sharma) [reply cmflib/commands/init/amazonS3.py:56]: Done.

File: cmflib/commands/init/amazonS3.py [hunk 26]
```diff
@@ -0,0 +1,125 @@
...
+                create_cmf_config(cmf_config, "http://127.0.0.1:80")
+        abs_path = None
+        file = "git_initialize.sh"
+        # find absolute path for git_initialize.sh
+        for root, dirs, files in os.walk(os.path.dirname(__file__)):
+            for name in files:
+                if name == file:
+                    abs_path = os.path.abspath(os.path.join(root, name))
+        # executing git_initialize.sh
+        result = subprocess.run(
```
[the remark this answers is not in the document]
Comment (varkha-d-sharma) [reply cmflib/commands/init/amazonS3.py:43]: Done.

File: cmflib/cli/utils.py [hunk 27]
```diff
@@ -1,4 +1,71 @@
...
+import os
+
+
 def fix_subparsers(subparsers):
     subparsers.required = True
     subparsers.dest = "cmd"
 
+
+def create_cmf_config(file_name: str, cmf_server_ip: str):
+    with open(file_name, "w") as file:
```
[the remark this answers is not in the document]
Comment (varkha-d-sharma) [reply cmflib/cli/utils.py]: Done

File: cmflib/dvc_config.py [hunk 28]
```diff
@@ -0,0 +1,72 @@
...
+    @staticmethod
+    def get_dvc_config():
+        result = subprocess.run(["dvc", "config", "-l"], capture_output=True, text=True)
```
[the remark this answers is not in the document]
Comment (varkha-d-sharma) [reply cmflib/dvc_config.py]: Done

File: cmflib/commands/artifact/push.py [hunk 29]
```diff
@@ -0,0 +1,36 @@
...
+                result = subprocess.run(["dvc", "push"], stdout=subprocess.PIPE, stderr=subprocess.STDOUT, text=True)
```
[the remark this answers is not in the document]
Comment (varkha-d-sharma) [reply cmflib/commands/artifact/push.py]: Done

File: cmflib/cli/utils.py [hunk 30]
```diff
@@ -1,4 +1,71 @@
...
+         file.write(f"cmf.server.ip={cmf_server_ip}")
+
+def read_cmf_config(file_name: str):
+    with open(file_name, "r") as file:
```
[the remark this answers is not in the document]
Comment (varkha-d-sharma) [reply cmflib/cli/utils.py]: Done

File: cmflib/commands/init/show.py (hunk 3 above)
Comment (varkha-d-sharma) [reply cmflib/commands/init/show.py]: Done

File: cmflib/commands/init/amazonS3.py [hunk 31]
```diff
@@ -0,0 +1,125 @@
...
+import subprocess
+import shlex
+
+from cmflib.cli.command import CmdBase
+from cmflib.cli.utils import create_cmf_config
+
+
+class CmdInitAmazonS3(CmdBase):
+    def run(self):
+        cmf_config = "./.cmfconfig"
```
[the remark this answers is not in the document]
Comment (varkha-d-sharma) [reply cmflib/commands/init/amazonS3.py]: Done

File: cmflib/commands/artifact/pull.py [hunk 32]
```diff
@@ -42,49 +70,100 @@ def run(self):
...
-            temp = name_url[1].split("/")
-            bucket_name = temp[2]
-            object_name = temp[3] + "/" + temp[4]
-            path_name = current_directory + "/" + name_url[0]
-            stmt = artifact_class_obj.download_artifacts(current_directory, bucket_name, object_name, path_name)
-            print(stmt)
+        dvc_config_op = dvc_config.get_dvc_config()  # pulling dvc config
+        if dvc_config_op[0] == "minio":  # if core.remote = minio
+            minio_class_obj = minio_artifacts.minio_artifacts()
+            for name_url in names_urls:
```
[the remark this answers is not in the document]
Comment (varkha-d-sharma) [reply cmflib/commands/artifact/pull.py:86]: Done

File: cmflib/commands/artifact/pull.py [hunk 33]
```diff
@@ -42,49 +70,100 @@ def run(self):
...
+                temp = name_url[1].split("/")
+                temp_var = temp[2].split(":")
+                host = temp_var[0]
+                download_loc = current_directory + "/" + name_url[0]
+                temp.pop(0)
+                temp.pop(0)
+                temp.pop(0)
+                current_loc_1 = "/".join(temp)
+                current_loc = f"/{current_loc_1}"
+                stmt = sshremote_class_obj.download_artifacts(
```
[the remark this answers is not in the document]
Comment (varkha-d-sharma) [reply cmflib/commands/artifact/pull.py:132]: Done

**Pull request merged**